### PR TITLE
[MRG] Added largest_connected_component_img

### PR DESCRIPTION
**Pull request by SRSteinkamp**
Work on Issue: #1296  
Created the largest_connected_component_img and fixed a small bug in largest_connected_component, where entering a nifti image returns True.

### CODE REVIEW

File: nilearn/_utils/ndimage.py [hunk 1]
```diff
@@ -38,6 +47,40 @@ def largest_connected_component(volume):
...
 
 
+def largest_connected_component_img(imgs):
+    """ Return the largest connected component of an image or list of images.
+
+    Parameters
+    ----------
+    imgs: Niimg-like object or iterable of Niimg-like objects
+        See http://nilearn.github.io/manipulating_images/input_output.html
+        Image(s) to extract the largest connected component from.
```
Comment (bthirion) [inline nilearn/_utils/ndimage.py]: I'm curious about what happens with a 4D image

File: nilearn/tests/test_ndimage.py [hunk 2]
```diff
@@ -21,6 +24,43 @@ def test_largest_cc():
...
+            assert_true(isinstance(out, list))
+            assert_true(len(out) == 2)
+            for o, s in zip(out, shapes):
+                assert_true(o.shape == (s))
+
+            # Single image as input
+            out = largest_connected_component_img(imgs[0])
+            assert_true(isinstance(out, Nifti1Image))
+            assert_true(out.shape == (shapes[0] ))
+
```
Comment (bthirion) [inline nilearn/tests/test_ndimage.py]: A test with a 4D image, then ?

**Comment by bthirion** [comment]
LGTM otherwise.

File: nilearn/_utils/ndimage.py (hunk 1 above)
Comment (SRSteinkamp) [inline nilearn/_utils/ndimage.py]: I had a short discussion with @KamalakerDadi  about this. We were a bit unsure about how ndimage would behave in the case of 4D images. I am going to look into this tonight.

**Comment by GaelVaroquaux** [comment]
> I had a short discussion with @KamalakerDadi about this. We were a bit unsure
> about how ndimage would behave in the case of 4D images. I am going to look
> into this tonight.

In case of a 4D image, ndimage is going to take the large CC in all
direction, ie in the 4D space. That is most likely going to surprise the
user. I think that, if the input image is 4D, we want to either raise an
error, or loop on every 3D image.

**Comment by bthirion** [comment]
I would loop on 3D images (and send a message to the user, at least in verbose mode).

**Comment by SRSteinkamp** [comment]
Ok, I'll implement the loop over the 3D images. I think this will be nicer from a user's perspective.

**Comment by AlexandreAbraham** [comment]
Note that for other functions we have chosen not to handle 3d and 4d in the same function. The reason is that we tried this previously and the codebase has grown to handle 3d/4d cases, and also users ended up feeding 4d images to some functions, thinking it was 3d, which resulted in unexpected results.

I think that we should add for 3d first, and discuss in a separate issue how to make it easy for users to process 4d images.

**Comment by GaelVaroquaux** [comment]
> I think that we should add for 3d first, and discuss in a separate issue how to make it easy for users to process 4d images.

It's my feeling too. The functionality is one that I would expect on 3D,
but less on 4D data.

**Comment by bthirion** [comment]
OK, let's forbid 4D images for the moment.

**Comment by SRSteinkamp** [comment]
So in the current form check_niimg_3d(img) should already provide the right error when a 4D image is given - at least if I got that right?

> I think that we should add for 3d first, and discuss in a separate issue how to make it easy for users to process 4d images.

For me (being more of a user): In my feeling there needs to be no added functionality as iter_image & concat_img should do the job already. But maybe adding an explicit example could do the trick, to stress the usefulness of those function (also new_img_like etc.). However, different issue.

**Comment by KamalakerDadi** [comment]
>So in the current form check_niimg_3d(img) should already provide the right error when a 4D image is given - at least if I got that right?

Yes.

File: nilearn/tests/test_ndimage.py [hunk 3]
```diff
@@ -21,6 +24,43 @@ def test_largest_cc():
     b[5, 5, 5] = 1
     np.testing.assert_equal(a, largest_connected_component(b))
 
+    img = testing.generate_labeled_regions(shape = (10,11,12),
+                                           n_regions = 2 )
+
+    assert_raises(ValueError, largest_connected_component, img)
```
Comment (KamalakerDadi) [inline nilearn/tests/test_ndimage.py]: Can you write a short comment here of what you are testing ?
It will be useful for anybody who can quickly follow and understand.

Same for both tests you have added here.

**Comment by KamalakerDadi** [comment]
Two things missing:

* Test with 4D image to check whether raises an error.

* Add this new feature in doc/whats_new.rst

File: nilearn/_utils/ndimage.py [hunk 4]
```diff
@@ -38,6 +47,40 @@ def largest_connected_component(volume):
     return labels == label_count.argmax()
 
 
+def largest_connected_component_img(imgs):
+    """ Return the largest connected component of an image or list of images.
+
```
Comment (KamalakerDadi) [inline nilearn/_utils/ndimage.py]: Can you add this here ?  ```.. versionadded:: 0.3``` with empty line space before and after.

In Nilearn, you can have a look at this link
https://github.com/nilearn/nilearn/blob/230279e885aae8f9ae28c99c291ed3b0b4e93588/nilearn/connectome/connectivity_matrices.py#L204

Ref for version added and other: http://www.sphinx-doc.org/en/stable/markup/para.html

**Comment by KamalakerDadi** [comment]
Some minor comments from my side.

**Comment by SRSteinkamp** [comment]
Added the suggestions. Sorry for the double commit.

File: doc/whats_new.rst [hunk 5]
```diff
@@ -100,6 +100,10 @@ Enhancements
     - Extensive plotting example shows how to use contours and filled contours
       on glass brain.
 
+    - Function :func:`nilearn._utils.ndimage.largest_connected_component_img`
```
Comment (KamalakerDadi) [inline doc/whats_new.rst]: >did this break the check ?

CircleCI build failed because reference you have given as ```nilearn._utils.ndimage.largest_connected_component_img``` could not be found. Since this function is not initialized in ```nilearn/_utils/__init__.py``` from ndimage module.

We could initialize this function in ```_utils/__init__.py``` from ndimage and refer as ``nilearn._utils.largest_connected_component_img``` directly in whats_new.rst. In this way we can find the target and build will go through.

File: doc/whats_new.rst [hunk 6]
```diff
@@ -100,6 +100,10 @@ Enhancements
...
+      so that niftis can now be passed to 
+      :func:`nilearn._utils.ndimage.largest_connected_component`.
```
Comment (KamalakerDadi) [inline doc/whats_new.rst]: Also, I could not understand why largest_connected_component is referred here. May be you could write "so that Nifti images can now be passed on directly to extract largest connected components in the images."

File: nilearn/_utils/ndimage.py [hunk 7]
```diff
@@ -38,6 +47,42 @@ def largest_connected_component(volume):
...
+    ..versionadded:: 0.3
```
Comment (KamalakerDadi) [inline nilearn/_utils/ndimage.py]: Should be ```.. versionadded:: 0.3```. Space is missing between ".." and "versionadded"

File: nilearn/tests/test_ndimage.py [hunk 8]
```diff
@@ -21,6 +25,48 @@ def test_largest_cc():
     b[5, 5, 5] = 1
     np.testing.assert_equal(a, largest_connected_component(b))
 
+    # Tests for correct errors, when an image or string are passed.
+    img = testing.generate_labeled_regions(shape = (10,11,12),
```
Comment (KamalakerDadi) [inline nilearn/tests/test_ndimage.py]: PEP: Missing white space after ",". Should be like shape=(10, 11, 12) and n_regions=2)

File: nilearn/tests/test_ndimage.py [hunk 9]
```diff
@@ -21,6 +25,48 @@ def test_largest_cc():
...
+
+
+def test_largest_cc_img():
+    """ Check the extraction of the largest connected component, for niftis
+
+    Similiar to smooth_img tests for largest connected_component_img, here also
+    only the added features for largest_connected_component are tested.
+    """
+
+    #Test whether dimension of 3Dimg and list of 3Dimgs are kept.
```
Comment (KamalakerDadi) [inline nilearn/tests/test_ndimage.py]: Space missing between "#" and Test

File: nilearn/tests/test_ndimage.py [hunk 10]
```diff
@@ -21,6 +25,48 @@ def test_largest_cc():
...
+    shapes = ((10, 11, 12), (13, 14, 15))
+    regions = [1,3]
```
Comment (KamalakerDadi) [inline nilearn/tests/test_ndimage.py]: PEP8: regions = [1, 3] space missing after ",".

File: nilearn/tests/test_ndimage.py [hunk 11]
```diff
@@ -21,6 +25,48 @@ def test_largest_cc():
...
+
+    img1 = testing.generate_labeled_regions(shape = shapes[0],
```
Comment (KamalakerDadi) [inline nilearn/tests/test_ndimage.py]: PEP8: Some unexpected whitespaces here. Should be like shape=shapes[0].

File: nilearn/tests/test_ndimage.py [hunk 12]
```diff
@@ -21,6 +25,48 @@ def test_largest_cc():
...
+                                            n_regions =regions[0] )
```
Comment (KamalakerDadi) [inline nilearn/tests/test_ndimage.py]: Here too, PEP8 unexpected white space between n_regions and = and space before ")"

File: nilearn/tests/test_ndimage.py [hunk 13]
```diff
@@ -21,6 +25,48 @@ def test_largest_cc():
...
+    img2 = testing.generate_labeled_regions(shape = shapes[1],
+                                            n_regions =regions[1] )
```
Comment (KamalakerDadi) [inline nilearn/tests/test_ndimage.py]: Same here PEP8 as commented above.

File: nilearn/tests/test_ndimage.py [hunk 14]
```diff
@@ -21,6 +25,48 @@ def test_largest_cc():
...
+            out = largest_connected_component_img(imgs)
+            assert_true(isinstance(out, list))
+            assert_true(len(out) == 2)
+            for o, s in zip(out, shapes):
+                assert_true(o.shape == (s))
+
+            # Single image as input
+            out = largest_connected_component_img(imgs[0])
+            assert_true(isinstance(out, Nifti1Image))
+            assert_true(out.shape == (shapes[0] ))
```
Comment (KamalakerDadi) [inline nilearn/tests/test_ndimage.py]: Unexpected space after [0]

File: nilearn/tests/test_ndimage.py [hunk 15]
```diff
@@ -21,6 +25,48 @@ def test_largest_cc():
...
+
+        # Test whether 4D Nifti throws the right error.
+        img_4D = testing.generate_fake_fmri(shapes[0], length =17)
```
Comment (KamalakerDadi) [inline nilearn/tests/test_ndimage.py]: Same here unexpected space.

Comment (KamalakerDadi) [inline nilearn/tests/test_ndimage.py]: flake8 package would be helpful to check these minor errors.

File: doc/whats_new.rst (hunk 5 above)
Comment (SRSteinkamp) [inline doc/whats_new.rst]: Would this mean adding 
`from .ndimage import largest_connected_component_img`  
and `'largest_connected_component_img'` to `__all__` in `_utils/__init__.py`?

(I'll do all the little changes later -- and I thought I spend extra attention to the little details :/ . I have flake8 installed though, but I am also quite new to vim - so apparently my setup still isn't working as expected.)

Comment (KamalakerDadi) [inline doc/whats_new.rst]: Yes. If you add then you can check by importing ```from nilearn._utils import largest_connected_component_img``` without necessarily specifying ```ndimage``` module.

**Comment by GaelVaroquaux** [comment]
> +    - Function :func:`nilearn._utils.ndimage.largest_connected_component_img`

I just realized: that function shouldn't be in nilearn._utils.ndimage,
which is a private utilities module, but in nilearn.image.

**Comment by KamalakerDadi** [comment]
Thanks for changes @SRSteinkamp . We have to wait because for some reason scipy server is down which leads to tests failed on CircleCI.

**Comment by KamalakerDadi** [comment]
Regarding CircleCI reference error, I missed to tell you that you need to add your function in ```doc/modules/reference.rst```. Should be in alphabetical order so right after ```load_img``` in mod:```nilearn.image```.
Can you add that please and hopefully CircleCI will be fixed ?

**Comment by SRSteinkamp** [comment]
If it's alphabetical, then it should be before load_img ;) 
Done, hopefully, everything goes through now. 
Thanks for guiding me through all this :)

File: nilearn/_utils/ndimage.py [hunk 16]
```diff
@@ -6,12 +6,12 @@
 
 import numpy as np
 from scipy import ndimage
-
-
```
Comment (KamalakerDadi) [inline nilearn/_utils/ndimage.py]: Ideally, it would be better to leave other things not to change which are not related to your PR. Sometimes reviewing PR might get complex. Just a suggestion. :)

**Comment by KamalakerDadi** [comment]
Thank you @SRSteinkamp for working on this. Feel free to shoot for another issue ;)

File: nilearn/_utils/ndimage.py (hunk 16 above)
Comment (SRSteinkamp) [inline nilearn/_utils/ndimage.py]: I see the point, won't do this in the future... I got a bit overzealous when I got flake working.

**Comment by SRSteinkamp** [comment]
Will do if I find some time at my hand :) 
Hm, I do not understand the CI error-message, but it seems to be related to the docs/sphinx again?

**Comment by KamalakerDadi** [comment]
>Hm, I do not understand the CI error-message, but it seems to be related to the docs/sphinx again?

Build is succeeded which your changes are fine. But, I don't get the error message.

**Comment by SRSteinkamp** [comment]
Ok ... Well I see "ci/circleci — Your tests failed on CircleCI ":

A small excerpt of the error message for build no. 2367:
``Exception occurred:
  File "/home/ubuntu/nilearn/doc/sphinxext/sphinx_gallery/docs_resolv.py", line 262, in _get_link
    raise``


``source continuous_integration/circle_ci_test_doc.sh returned exit code 2``

**Comment by KamalakerDadi** [comment]
>Ok ... Well I see "ci/circleci — Your tests failed on CircleCI ":

I see that but I don't know what is causing ?
For me your PR looks good.

**Comment by KamalakerDadi** [comment]
All went green. Can I have reviews or should I merge this one ?
Please let me know if I missed to review ?

File: nilearn/_utils/ndimage.py [hunk 17]
```diff
@@ -25,6 +25,11 @@ def largest_connected_component(volume):
     volume: numpy.array
         3D boolean array with only one connected component.
     """
+    if hasattr(volume, "get_data") \
+       or isinstance(volume, _basestring):
```
Comment (GaelVaroquaux) [inline nilearn/_utils/ndimage.py]: Style: Python developers do not like continuation lines. We prefer to use parenthesis. We will merge this, to avoid more waiting time, but change it after merge.

File: nilearn/image/image.py [hunk 18]
```diff
@@ -901,3 +901,41 @@ def load_img(img, wildcards=True, dtype=None):
         that the returned object has get_data() and affine attributes.
     """
     return check_niimg(img, wildcards=wildcards, dtype=dtype)
+
+
+def largest_connected_component_img(imgs):
+    """ Return the largest connected component of an image or list of images.
+
+    .. versionadded:: 0.3
```
Comment (GaelVaroquaux) [inline nilearn/image/image.py]: Need to change this to 0.3.1. We'll do it in master

**Comment by GaelVaroquaux** [comment]
Merging. Thanks a lot
